consensus/src/dag/tests/order_rule_tests.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    dag::{
+        anchor_election::RoundRobinAnchorElection,
+        dag_store::Dag,
+        order_rule::OrderRule,
+        tests::{dag_test::MockStorage, helpers::new_certified_node},
+        types::NodeCertificate,
+        CertifiedNode,
+    },
+    test_utils::placeholder_ledger_info,
+};
+use aptos_consensus_types::common::Author;
+use aptos_infallible::RwLock;
+use aptos_types::{
+    aggregate_signature::AggregateSignature, epoch_state::EpochState,
+    validator_verifier::random_validator_verifier,
+};
+use futures_channel::mpsc::{unbounded, UnboundedReceiver};
+use proptest::prelude::*;
+use std::sync::Arc;
+
+/// Generate a virtual dag that first layer represents round
+/// second layer represents nodes
+/// third layer is a bitmask that represents strong links (true => linked, false => not linked)
+fn generate_virtual_dag(
+    num_validators: usize,
+    round: u64,
+) -> impl Strategy<Value = Vec<Vec<Vec<bool>>>> {
+    let f = (num_validators - 1) / 3;
+    let bitmask: Vec<bool> = std::iter::repeat(false)
+        .take(f)
+        .chain(std::iter::repeat(true).take(num_validators - f))
+        .collect();
+    proptest::collection::vec(
+        proptest::collection::vec(Just(bitmask).prop_shuffle(), num_validators),
+        round as usize,
+    )
+}
+
+/// Generate `num_per` random permutations of how nodes are processed by the order rule
```

Comment:
possible elaborate this comment? IIUC, this basically picks X node indices out of a total Y nodes and these indices are used as the order in which the node is added to order rule, correct?

Reply:
it's just a permutation of the order of all Y nodes being added